### PR TITLE
Apply Commission only on gain and not on whole stake

### LINKED ISSUE #301: Commission is calculated on total stake and not profit
**Describe the bug**
Currently, the commission is calculated on total stake. This is not ideal, eg., if commission is 10% and you  stake 1 million, you automatically lose 100k.

**Expected behavior**
Commission should be set on profits and not on stake

### LINKED ISSUE #368: In set commission, there is no upperlimit
As of now 
There is no upper limit on commission in setCommission
https://github.com/razor-network/contracts/blob/ac97f68a50813ef496a9d1f17514ce3cf06cf59c/contracts/Core/StakeManager.sol#L225

Exploit could happen here due to this, like setting commission 10000, and running away in withdraw
https://github.com/razor-network/contracts/blob/ac97f68a50813ef496a9d1f17514ce3cf06cf59c/contracts/Core/StakeManager.sol#L196

But we were saved by this line
https://github.com/razor-network/contracts/blob/ac97f68a50813ef496a9d1f17514ce3cf06cf59c/contracts/Core/StakeManager.sol#L198
if commission exceeds rAmount, it would have reverted so there is no exploit

now as we have moved to gain in this PR #367 this issue becomes significant
as commission on the gain by 1000 % may not exceed rAmount but will exceed gain, and there is no check on that

hence we need to introduce max limit on commission

### PR DESCRIPTION
**Pull request by abhishekvispute**
Fixes #301 
```
Basically Have RazorPutIn as param, which keeps record of how many razor, anyone has put in.
Then at anytime we can calculate gain on X amount of sRZR by

X * (Current Releation) - X * (RazorPutIn/ TotalSupplysRZR)


Here  (RazorPutIn/ TotalSupplysRZR) signifies for 1 sRZR user hold, how much they had put in

```
Tests are updated for 3 cases 
gain<=0  :  no commission
gain > 0  :  commission on gain

have added one scenario test case as well
it stimulates this
https://docs.google.com/spreadsheets/d/1b8ks98mRczDIX9tayjgCxI5NvD7Hq27JSYVWyqCfXmg/edit?usp=sharing

Now as we moved to commission on gain
It created this new issue #368 
Hence have fixed it as well here, basically have set max cap on commission
Fixes #368

### CODE REVIEW

File: contracts/tokenization/StakedToken.sol [hunk 1]
```diff
@@ -24,4 +33,39 @@ contract StakedToken is ERC20 {
...
     }
+
+    /// @notice Used in stake, delegate
+    /// Purpose is to maitain quotient
+    function updateQuotient(
+        address user,
+        uint256 numAddition,
+        uint256 denAddition
+    ) external onlyOwner returns (bool) {
+        quotient[user].numerator = quotient[user].numerator + numAddition;
```
[the remark this answers is not in the document]
Comment (abhishekvispute) [reply contracts/tokenization/StakedToken.sol]: done https://github.com/razor-network/contracts/pull/367/commits/5e86e1becce0e0fbabe83afede13dcf8bfee69fa

[the remark this answers is not in the document]
Comment (abhishekvispute) [reply contracts/tokenization/StakedToken.sol]: now there is no quotient 
We only maintain how much razor, user has put in

File: contracts/Core/StakeManager.sol [hunk 2]
```diff
@@ -71,23 +71,22 @@ contract StakeManager is Initializable, ACL, StakeStorage, StateManager, Pause {
...
+            stakers[numStakers] = Structs.Staker(false, 0, msg.sender, address(sToken), numStakers, 10000, epoch, amount);
             stakerId = numStakers;
             stakerIds[msg.sender] = stakerId;
         } else {
             IStakedToken sToken = IStakedToken(stakers[stakerId].tokenAddress);
             uint256 totalSupply = sToken.totalSupply();
             uint256 toMint = _convertRZRtoSRZR(amount, stakers[stakerId].stake, totalSupply); // RZRs to sRZRs
-
+            // Mint sToken as Amount * (totalSupplyOfToken/previousStake)
+            sToken.mint(msg.sender, toMint);
```
Comment (abhishekvispute) [inline contracts/Core/StakeManager.sol:86]: mint is moved up to stake update, as in mint we calculate razorPutIn, which is proportional to currentStake

File: contracts/Core/StakeManager.sol [hunk 3]
```diff
@@ -186,18 +185,22 @@ contract StakeManager is Initializable, ACL, StakeStorage, StateManager, Pause {
...
         if (stakerIds[msg.sender] != stakerId && staker.commission > 0) {
-            uint256 commission = (rAmount * staker.commission) / 100;
-            require(razor.transfer(staker._address, commission), "couldnt transfer");
-            rAmount = rAmount - commission;
+            // Calculate Gain
+            uint256 initial = sToken.getRZRPutIn(msg.sender, lock.amount);
+            if (rAmount > initial) {
+                uint256 gain = rAmount - initial;
+                uint256 commission = (gain * staker.commission) / 100;
+                rAmount = rAmount - commission;
```
Comment (hrishikeshio) [inline contracts/Core/StakeManager.sol:196]: what if commission > rAmount?

Comment (hrishikeshio) [reply contracts/Core/StakeManager.sol:196]: nvm its not possible.

File: contracts/tokenization/IStakedToken.sol [hunk 4]
```diff
@@ -31,4 +31,6 @@ interface IStakedToken is IERC20 {
      * - `account` must have at least `amount` tokens.
      */
     function burn(address account, uint256 amount) external returns (bool);
+
+    function getRZRPutIn(address delegator, uint256 sAmount) external view returns (uint256);
```
Comment (hrishikeshio) [inline contracts/tokenization/IStakedToken.sol:35]: Instead of PutIn you can use Deposited.

Comment (abhishekvispute) [reply contracts/tokenization/IStakedToken.sol]: done https://github.com/razor-network/contracts/pull/367/commits/1f97e6752af7f94b48ee399e588abd13ce773033

**Review by hrishikeshio — APPROVED** [review]

**Review by hrishikeshio — DISMISSED** [review]

**Comment by abhishekvispute** [comment]
Hi Guys , @hrishikeshio , @SkandaBhat , @SamAg19 
Due to recent merged PR of change in lock mech. https://github.com/razor-network/contracts/pull/374, we had to do some changes in this commission PR as well.

In https://github.com/razor-network/contracts/pull/374, the amount was locked in RZR only at the time of unstake.
Due to this, we have to move commission calculation to unstake as well.
As the gain is calculated from sRZRS and it should be calculated at that RZR valuation only.

**Now, we calculate commission in unstake, we store it in lock, then in withdrawing we transfer**

File: contracts/Core/StakeManager.sol (hunk 2 above)
Comment (hrishikeshio) [reply contracts/Core/StakeManager.sol:86]: external contract interaction before state update. re-entrancy vuln.

File: contracts/Core/StakeManager.sol [hunk 5]
```diff
@@ -111,12 +110,12 @@ contract StakeManager is Initializable, ACL, StakeStorage, StateManager, Pause {
...
-        stakers[stakerId].stake = stakers[stakerId].stake + (amount);
-
-        // Step 4:  Mint sToken as Amount * (totalSupplyOfToken/previousStake)
+        // Step 3:  Mint sToken as Amount * (totalSupplyOfToken/previousStake)
         sToken.mint(msg.sender, toMint);
 
+        // Step 4: Increase given stakers stake by : Amount
+        stakers[stakerId].stake = stakers[stakerId].stake + (amount);
+
         emit Delegated(msg.sender, epoch, stakerId, stakers[stakerId].stake, block.timestamp);
```
Comment (hrishikeshio) [inline contracts/Core/StakeManager.sol:119]: needs to be moved up

Comment (abhishekvispute) [reply contracts/Core/StakeManager.sol:119]: I think this would be solved once we merge adtiyas PR

File: contracts/Core/StakeManager.sol [hunk 6]
```diff
@@ -181,23 +191,20 @@ contract StakeManager is Initializable, ACL, StakeStorage, StateManager, Pause {
...
-            require(razor.transfer(staker._address, commission), "couldnt transfer");
+        if (lock.commission > 0) {
+            withdrawAmount = lock.amount - lock.commission;
+            require(razor.transfer(staker._address, lock.commission), "couldnt transfer");
         }
 
-        //Transfer stake
-        require(razor.transfer(msg.sender, lockedAmount), "couldnt transfer");
+        // Reset lock
+        _resetLock(stakerId);
```
Comment (hrishikeshio) [inline contracts/Core/StakeManager.sol:202]: state change after transfer. this should be moved up

File: contracts/Core/StakeManager.sol (hunk 5 above)
Comment (abhishekvispute) [reply contracts/Core/StakeManager.sol:119]: oh no we can't move this up 
mint has to happen before stake increase as there is dependency

when we mint, in beforeTokenTransfer, we calculate razor deposited 
their we need an old stake, not the new one 

https://github.com/abhishekvispute/contracts/blob/52ec788e54fa8d320f0863944173485675948aa9/contracts/tokenization/StakedToken.sol#L63

File: contracts/tokenization/StakedToken.sol [hunk 7]
```diff
@@ -2,26 +2,79 @@
...
     }
 
-    function burn(address account, uint256 amount) external onlyOwner returns (bool) {
+    function burn(address account, uint256 amount) external override onlyOwner returns (bool) {
         _burn(account, amount);
         return true;
     }
+
+    /// @notice Used in withdraw
+    // At any time via calling this one can find out how much RZR was invested for this much sRZR
```
Comment (hrishikeshio) [inline contracts/tokenization/StakedToken.sol:44]: remove 'invested'

File: contracts/Core/StakeManager.sol (hunk 2 above)
Comment (abhishekvispute) [reply contracts/Core/StakeManager.sol:86]: we can't move this up
mint has to happen before stake increase as there is dependency

when we mint, in beforeTokenTransfer, we calculate razor deposited
their we need an old stake, not the new one

https://github.com/abhishekvispute/contracts/blob/52ec788e54fa8d320f0863944173485675948aa9/contracts/tokenization/StakedToken.sol#L63

File: contracts/Core/Parameters.sol [hunk 8]
```diff
@@ -101,6 +101,12 @@ contract Parameters is ACL, Constants {
         maxAge = _maxAge;
     }
 
+    function setMaxCommission(uint8 _maxCommission) external onlyRole(DEFAULT_ADMIN_ROLE) {
```
[the remark this answers is not in the document]
Comment (abhishekvispute) [reply contracts/Core/Parameters.sol:104]: resolved 
https://github.com/abhishekvispute/contracts/blob/74bb27c6e26dd2c66a203e9469ec3b81531eae95/contracts/Core/StakeManager.sol#L159

File: contracts/Core/StakeManager.sol (hunk 5 above)
Comment (abhishekvispute) [reply contracts/Core/StakeManager.sol]: resolved 
As we discussed modified mint directly. 
https://github.com/abhishekvispute/contracts/blob/74bb27c6e26dd2c66a203e9469ec3b81531eae95/contracts/tokenization/StakedToken.sol#L37

File: contracts/Core/StakeManager.sol (hunk 6 above)
Comment (abhishekvispute) [reply contracts/Core/StakeManager.sol]: resolved

File: contracts/tokenization/StakedToken.sol (hunk 7 above)
Comment (abhishekvispute) [reply contracts/tokenization/StakedToken.sol]: resolved

File: contracts/Core/StakeManager.sol (hunk 2 above)
Comment (abhishekvispute) [reply contracts/Core/StakeManager.sol]: resolved

**Review by hrishikeshio — APPROVED** [review]

**Review by SamAg19 — APPROVED** [review]

**Review by hrishikeshio — APPROVED** [review]

**Pull request merged**